drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * complete scene for each time the user moves its cursor, we can update only
+ * the cursor plane and it will be automatically overlayed by the hardware on
+ * top of the primary plane. There's no need of software composition in this
+ * case.
+ *
+ * But there was synchronisation problems related to multiple planes
+ * usage. The KMS API was not atomic, so you'd have to update the primary
+ * plane and then the overlay planes with distinct IOCTL's. This could lead
+ * to tearing and also some trouble related to blocking, so the atomic
+ * API was proposed to fix these problems.
+ * 
+ * With the introduction of the KMS atomic API, all the planes could get
+ * updated in a single IOCTL, using drmModeAtomicCommit(). This can be
+ * either asynchronous or fully blocking.
+ *
+ * This example assumes that you are familiar with modeset-vsync. Only
+ * the differences between both files are highlighted here.
+ */
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <time.h>
+#include <unistd.h>
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+#include <drm_fourcc.h>
+
+struct modeset_buf;
+struct modeset_dev;
+static int modeset_find_plane(int fd, struct modeset_dev *dev);
+static int modeset_atomic_modeset(int fd, struct modeset_dev *dev);
+static int modeset_find_crtc(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev);
+static int modeset_create_fb(int fd, struct modeset_buf *buf);
+static void modeset_destroy_fb(int fd, struct modeset_buf *buf);
+static int modeset_setup_dev(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev);
+static int modeset_open(int *out, const char *node);
+static int modeset_prepare(int fd);
+static void modeset_draw(int fd);
+static void modeset_draw_dev(int fd, struct modeset_dev *dev);
+static void modeset_cleanup(int fd);
+
+/*
+ * A new struct is introduced: drm_object. It stores properties of certain
+ * objects (connectors, CRTC and planes) that are used in atomic modeset setup
+ * and also in atomic page-flips (all planes updated in a single IOCTL).
+ */
+
+struct drm_object {
+	drmModeObjectProperties *props;
+	drmModePropertyRes **props_info;
+};
+
+struct modeset_buf {
+	uint32_t width;
+	uint32_t height;
+	uint32_t stride;
+	uint32_t size;
+	uint32_t handle;
+	uint8_t *map;
+	uint32_t fb;
+};
+
+struct modeset_dev {
+	struct modeset_dev *next;
+
+	unsigned int front_buf;
+	struct modeset_buf bufs[2];
+
+	struct drm_object connector;
+	struct drm_object crtc;
+	struct drm_object plane;
+
+	drmModeModeInfo mode;
+	uint32_t plane_id;
+	uint32_t conn_id;
+	uint32_t crtc_id;
+	uint32_t crtc_index; /* this is not really persistent */
+	drmModeCrtc *saved_crtc;
+
+	bool pflip_pending;
+	bool cleanup;
+
+	uint8_t r, g, b;
+	bool r_up, g_up, b_up;
+};
+static struct modeset_dev *modeset_list = NULL;
```

Comment:
So this would actually be an output list.

Reply:
Done.